python/ray/spark/utils.py

```diff
@@ -0,0 +1,411 @@
+            "create a spark session and connect to a spark cluster."
+        )
+    return spark_session
+
+
+def get_spark_application_driver_host(spark):
+    return spark.conf.get("spark.driver.host")
+
+
+def get_max_num_concurrent_tasks(spark_context):
+    """Gets the current max number of concurrent tasks."""
+    # pylint: disable=protected-access
+    # spark version 3.1 and above have a different API for fetching max concurrent tasks
+    if spark_context._jsc.sc().version() >= "3.1":
+        return spark_context._jsc.sc().maxNumConcurrentTasks(
+            spark_context._jsc.sc().resourceProfileManager().resourceProfileFromId(0)
+        )
+    return spark_context._jsc.sc().maxNumConcurrentTasks()
+
+
+def _get_total_physical_memory():
+    import psutil
+
+    return psutil.virtual_memory().total
+
+
+def _get_total_shared_memory():
+    import shutil
+
+    return shutil.disk_usage("/dev/shm").total
+
+
+def _get_cpu_cores():
+    import multiprocessing
+
+    return multiprocessing.cpu_count()
+
+
+def _calc_mem_per_ray_worker(
+    num_task_slots, physical_mem_bytes, shared_mem_bytes, heap_to_object_store_ratio
+):
+    available_physical_mem_per_node = int(
+        physical_mem_bytes / num_task_slots * _MEMORY_BUFFER_OFFSET
+    )
+    available_shared_mem_per_node = int(
+        shared_mem_bytes / num_task_slots * _MEMORY_BUFFER_OFFSET
+    )
+
+    ray_worker_object_store_bytes = int(
+        min(
+            available_physical_mem_per_node * heap_to_object_store_ratio,
+            available_shared_mem_per_node,
+        )
+    )
+    ray_worker_heap_mem_bytes = (
+        available_physical_mem_per_node - ray_worker_object_store_bytes
+    )
+    return ray_worker_heap_mem_bytes, ray_worker_object_store_bytes
+
+
+def _resolve_target_spark_tasks(calculated_limits):
+    """
+    Return the max value of a list of spark task total count calculations based on the
+    provided configuration arguments to `init_cluster`.
+    Args:
+        calculated_limits: A list of calculated values wherein the highest value based on
+            spark cluster worker instance sizes and user-specified ray cluster configuration is
+            taken to ensure that spark cluster limits are not exceeded.
+
+    Returns: The maximum calculated number of spark tasks for the configured ray cluster.
+
+    """
+    return max(*calculated_limits)
+
+
+def get_target_spark_tasks(
+    max_concurrent_tasks,
+    num_spark_task_cpus,
+    num_spark_task_gpus,
+    ray_worker_heap_memory_bytes,
+    ray_worker_object_store_memory_bytes,
+    num_spark_tasks,
+    total_cpus,
+    total_gpus,
+    total_heap_memory_bytes,
+    total_object_store_memory_bytes,
+):
+
+    if num_spark_tasks is not None:
+        if num_spark_tasks == -1:
+            # num_spark_tasks=-1 represents using all available spark task slots
+            num_spark_tasks = max_concurrent_tasks
+        elif num_spark_tasks <= 0:
```

Comment:
Note: now we already switch to use normal spark job instead of spark barrier mode job. So now we can allow `num_spark_tasks` argument larger than `max_concurrent_tasks`, in this case, it will trigger spark cluster scale up (if you config spark to be autoscaling)